crates/bevy_ecs/src/system/combinator.rs

```diff
@@ -316,3 +316,69 @@ where
         b(value)
     }
 }
+
+/// A [`System`] created by piping the unwrapped output of the first system into the input of the second.
+/// If the output of the first system is [`None`], the second system is not executed.
+///
+/// This can be repeated indefinitely. If not [`None`], the unwrapped output is consumed by the receiving system.
+///
+/// Given two systems `A` and `B`, A may be connected with `B` as `A.pipe_map(B)` if the output type of `A` is
+/// an [`Option`] of the input type of `B`. `B` must also return a type that implements [`Default`].
+///
+/// Note that for [`FunctionSystem`](crate::system::FunctionSystem)s the output is the return value
+/// of the function and the input is the first [`SystemParam`](crate::system::SystemParam) if it is
+/// tagged with [`In`](crate::system::In) or `()` if the function has no designated input parameter.
+///
+/// # Examples
+///
+/// ```
+/// use std::num::ParseIntError;
+///
+/// use bevy_ecs::prelude::*;
+///
+/// fn main() {
+///     let mut world = World::default();
+///     // pipe_map_some the `parse_message_system`'s output into the `check_system`s input
+///     let mut pipe_map_some_system = parse_message_system.pipe_map_some(triple);
+///     pipe_map_some_system.initialize(&mut world);
+///
+///     world.insert_resource(Message("not a usize".to_string()));
+///     assert_eq!(pipe_map_some_system.run((), &mut world), 0);
+///
+///     world.insert_resource(Message("14".to_string()));
+///     assert_eq!(pipe_map_some_system.run((), &mut world), 42);
+/// }
+///
+/// #[derive(Resource)]
+/// struct Message(String);
+///
+/// fn parse_message_system(message: Res<Message>) -> Option<usize> {
+///     message.0.parse::<usize>().ok()
+/// }
+///
+/// fn triple(In(parsed): In<usize>) -> usize {
+///     parsed * 3
+/// }
+/// ```
+pub type PipeMapSomeSystem<SystemA, SystemB> = CombinatorSystem<PipeMapSome, SystemA, SystemB>;
+
+#[doc(hidden)]
+pub struct PipeMapSome;
+
+impl<A, B> Combine<A, B> for PipeMapSome
+where
+    A: System<Out = Option<B::In>>,
+    B: System,
+    B::Out: Default,
+{
+    type In = A::In;
+    type Out = B::Out;
+
+    fn combine(
+        input: Self::In,
+        a: impl FnOnce(A::In) -> Option<B::In>,
+        b: impl FnOnce(B::In) -> B::Out,
+    ) -> Self::Out {
+        a(input).map(b).unwrap_or_default()
```

Comment:
Why not just return `Option<B::Out>` and remove the `B::Out: Default` requirement? To replicate the current behaviour, an end user could then call `sys_a.pipe_map_some(sys_b).map(Option::unwrap_or_default)`.